File: c++/494. Target Sum.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int S) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (sum < S || (S + sum)&1) return 0;

        int w = (S + sum) / 2;
        vector<int> dp(w+1, 0);
        dp[0] = 1;
        for (int num : nums) {
            for (int i = w; i >= num; i--) {
                dp[i] += dp[i-num];
            }
        }

        return dp[w];
    }
};

    /**
     * 494
     * 输入: nums: [1, 1, 1, 1, 1], S: 3
     * 输出: 5
     * 解释:
     * -1+1+1+1+1 = 3
     * +1-1+1+1+1 = 3
     * +1+1-1+1+1 = 3
     * +1+1+1-1+1 = 3
     * +1+1+1+1-1 = 3
     * 
     * sum(P) 前面符号为+的集合；sum(N) 前面符号为减号的集合
     * 所以题目可以转化为
     * sum(P) - sum(N) = target 
     * => sum(nums) + sum(P) - sum(N) = target + sum(nums)
     * => 2 * sum(P) = target + sum(nums) 
     * => sum(P) = (target + sum(nums)) / 2
     * 因此题目转化为01背包，也就是能组合成容量为sum(P)的方式有多少种
     */

File: go/src/典型问题/DP/494. Target Sum/494. Target Sum.go
package main

import "fmt"

/**
https://leetcode-cn.com/problems/target-sum/
思路：根据题设可以讲数组分为正子集P和负子集N, P-N=S, P+N=T, T为数组所有元素之和
可以推出P = (T+S)/2, 题目转化为从数组中选取若干个元素，使得和为(T+S)/2
 */
func findTargetSumWays(nums []int, target int) int {
	var s int
	for _, num := range nums {
		s += num
	}
	if s < target || (s + target) & 1 == 1 {
		return 0
	}
	var t = (s + target)/2
	dp := make([]int, t+1)
	dp[0] = 1
	for _, num := range nums {
		for i := t; i >= num; i-- {
			dp[i] += dp[i-num]
		}
	}
	return dp[t]
}

func main() {
	fmt.Println(findTargetSumWays([]int{1,1,1,1,1}, 3))
	fmt.Println(findTargetSumWays([]int{1}, 1))
}
